pkg/lightning/backend/tidb/tidb.go

```diff
@@ -411,27 +482,54 @@ func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, colu
 		insertStmt.WriteByte(')')
 	}
 	insertStmt.WriteString(" VALUES")
+	return &insertStmt
+}
 
-	// Note: we are not going to do interpolation (prepared statements) to avoid
-	// complication arise from data length overflow of BIT and BINARY columns
+func (be *tidbBackend) execStmts(ctx context.Context, stmtTasks []stmtTask, batch bool) error {
+	for _, stmtTask := range stmtTasks {
+		for i := 0; i < writeRowsMaxRetryTimes; i++ {
+			stmt := stmtTask.stmt
+			_, err := be.db.ExecContext(ctx, stmt)
+
+			failpoint.Inject("mockNonRetryableError", func() {
+				maxErrorCount = 3
+				// To mock the non-retryable error for TestWriteRowsErrorSkip test, we will fail the execStmts when:
+				//   1. It's in batch mode.
+				//   2. It's inserting the row ("1").
+				//   3. It's inserting the row ("3").
+				if batch || stmtTask.rows[0] == tidbRow("1") || stmtTask.rows[0] == tidbRow("3") {
+					err = stderrors.New("mock non-retryable error")
+				}
+			})
 
-	for i, row := range rows {
-		if i != 0 {
-			insertStmt.WriteByte(',')
+			if err != nil {
+				if !common.IsContextCanceledError(err) {
+					log.L().Error("execute statement failed",
+						zap.Array("rows", stmtTask.rows), zap.String("stmt", redact.String(stmt)), zap.Error(err))
+				}
+				// It's batch mode, just return the error.
+				if batch {
+					return errors.Trace(err)
+				}
+				// Retry the non-batch insert here.
+				if common.IsRetryableError(err) {
+					continue
+				}
+				// TODO: record the error.
+				// Check the error count.
+				be.curErrorCount.Inc()
+				if be.curErrorCount.Load() >= uint64(maxErrorCount) {
+					return errors.Trace(err)
+				}
```

Comment:
i mean just remove the entire error counting logic (only keep the TODO comment). the max-error implementation will not use a localized error count.

Reply:
Do you mean just keep the downgrading logic in this PR🤔?

Reply:
yes, just keep the PR focusing on downgrading.

Reply:
Done. PTAL.